### ISSUE TITLE
Add sig 9 kill command

### ISSUE THREAD

**Issue by simonhughxyz** [riverwm/river#984]
It would be nice to add a command that send a sig 9 `kill` to a focused window.
Sometimes windows freeze or misbehave and the `close` command is not enough.

**Comment by ifreund** [comment]
I don't think this is in scope for river. I think something like htop is the correct tool for the use-case of killing a buggy program.

**Comment by simonhughxyz** [comment]
I disagree, the way I see it a window managers purpose is supposed to manage windows, sometimes windows do not respond and won't close, if a windows manager does not have the correct tools to do one of its jobs, i.e. close windows, then that is a missing bug or a missing feature.

If you won't include it naively, at least maybe include a way for people to script it. Is there a way to  return the process ID of a focused window?

**Comment by Leon-Plickat** [comment]
There is no good / reliable / unspoofable way to get a PID for a wayland client.

> [...] the way I see it a window managers purpose is supposed to manage windows, sometimes windows do not respond and won't close [...]

A window not responding is a client bug. It is not the job of river to work around client bugs.

---

Right now river asks the window to close, which is the correct thing for all well-behaved, reasonable clients.

We could, instead of killing the PID, forcefully disconnect the client (This is also what xkill does, btw, since you can't reliably map PID <-> Window on X either). Thoughts @ifreund ?

**Comment by ifreund** [comment]
Yeah, I definitely don't want to expose flaky client PID info from river.

I guess a command to drop a wayland client would be fine with me, it just needs to be well documented. Also we need a good answer for how it works with xwayland clients.

This really isn't a use-case I've ever had myself though to be honest, what programs are regularly hanging like this?

**Comment by simonhughxyz** [comment]
I am currently having problems with mpv, after a while some of them freeze and become unresponsive, a sig 9 is the only way to kill them.
If it helps, I've been transitioning from dwl to river and I do not have this problem under dwl.
So maybe it is a river bug.

**Comment by ifreund** [comment]
A WAYLAND_DEBUG=1 log of that freeze happening would probably be interesting.

**Comment by simonhughxyz** [comment]
I captures the log of a mpv that froze, here is the tail of that log:
```
[1717638.441] wl_surface@8.leave(wl_output@24)
[1717638.459] wl_surface@8.enter(wl_output@24)
[1717642.288] wl_surface@8.leave(wl_output@24)
[1717642.305] wl_surface@8.enter(wl_output@24)
[1717646.118] wl_surface@8.leave(wl_output@24)
[1717646.135] wl_surface@8.enter(wl_output@24)
[1717649.963] wl_surface@8.leave(wl_output@24)
[1717649.980] wl_surface@8.enter(wl_output@24)
file descriptor expected, object (40), message format_table(hu)
Killed
```

**Comment by ifreund** [comment]
> I captures the log of a mpv that froze, here is the tail of that log:

Interesting, could you perhaps obtain a `river -log-level debug` log as well? And share the full log please?

**Comment by ifreund** [comment]
Hmm, there doesn't seem to be anything relevant in the compositor logs. I assume these logs are from the point in time where mpv starts to hang? It would also be useful to see the full mpv log.

I forgot to confirm, you are running river's master branch right? If so I suspect this patch may work around the issue which may be on river's side or mpv's side, it's hard to tell:
```diff
diff --git a/river/Root.zig b/river/Root.zig
index bb89239..b023b0f 100644
--- a/river/Root.zig
+++ b/river/Root.zig
@@ -120,8 +120,6 @@ pub fn init(self: *Self) !void {
     const scene = try wlr.Scene.create();
     errdefer scene.tree.node.destroy();
 
-    scene.setLinuxDmabufV1(server.linux_dmabuf);
-
     const interactive_content = try scene.tree.createSceneTree();
     const drag_icons = try scene.tree.createSceneTree();
     const hidden_tree = try scene.tree.createSceneTree();
```
It would be useful to know if that patch affects things, the wlroots scene graph finely-grained dmabuf feedback implementation is quite spammy and I suspect it may be leading to resource exhaustion on mpv's side.

**Comment by novakne** [comment]
> I forgot to confirm, you are running river's master branch right? 

> error(xdg_shell): client is buggy and initiated size change while tiled or fullscreen

It seems to be it's patched with at least #924 too

**Comment by simonhughxyz** [comment]
> Hmm, there doesn't seem to be anything relevant in the compositor logs. I assume these logs are from the point in time where mpv starts to hang? It would also be useful to see the full mpv log.
> 
> I forgot to confirm, you are running river's master branch right? If so I suspect this patch may work around the issue which may be on river's side or mpv's side, it's hard to tell:
> 
> ```diff
> diff --git a/river/Root.zig b/river/Root.zig
> index bb89239..b023b0f 100644
> --- a/river/Root.zig
> +++ b/river/Root.zig
> @@ -120,8 +120,6 @@ pub fn init(self: *Self) !void {
>      const scene = try wlr.Scene.create();
>      errdefer scene.tree.node.destroy();
>  
> -    scene.setLinuxDmabufV1(server.linux_dmabuf);
> -
>      const interactive_content = try scene.tree.createSceneTree();
>      const drag_icons = try scene.tree.createSceneTree();
>      const hidden_tree = try scene.tree.createSceneTree();
> ```
> 
> It would be useful to know if that patch affects things, the wlroots scene graph finely-grained dmabuf feedback implementation is quite spammy and I suspect it may be leading to resource exhaustion on mpv's side.

I did try this patch for the past 3 days, and so far it seems to have fixed the problem.

**Comment by ifreund** [comment]
> I did try this patch for the past 3 days, and so far it seems to have fixed the problem.

Interesting, thanks for testing. I've asked around a bit on IRC and it seems like this upstream libwayland bug report may be relevant: https://gitlab.freedesktop.org/wayland/wayland/-/issues/422

At the same time, I haven't ever seen this error happen in the wild before and it seems to me that wlroots shouldn't really be sending enough fds to the client to trigger it. There seems to be potential for a bugfix/rate limiting on the wlroots side as well.

**Comment by ifreund** [comment]
I decided to just disable this apparently overly spammy feedback for now, it can be re-enabled at some point in the future with upstream improvements: 12e6833f316a56e3dc4d31a4708ae3b0bbea7070
